File: client/src/catalog/view/library.rs
//! The library landing (oracle: `LibraryPage.scala`, final post-33 design): the hero pill +
//! the guided-tour carousel + CTAs, then the book grid — category bands and cards linking
//! to each book's first lesson. The hero renders before the catalog loads; the grid waits.

use leptos::prelude::*;
use synapse_shared::catalog::{BookDto, CatalogEntryDto, SynapseIndexDto};

use crate::api::AsyncResult;
use crate::catalog::logic;
use crate::catalog::state;
use crate::catalog::view::tour::SynapseTour;

#[component]
pub fn LibraryPage() -> impl IntoView {
    let index = state::CatalogStore::from_context().index();
    let index_opt: Signal<Option<SynapseIndexDto>> = Signal::derive(move || match index.get() {
        AsyncResult::Loaded(idx) => Some(idx),
        _ => None,
    });
    view! {
        <div class="library">
            <section class="lib-hero">
                <div class="lib-hero__pill">
                    <span class="lib-hero__dot" aria-hidden="true"></span>
                    "A guided tour — everything Synapse can do"
                </div>
                <SynapseTour index=index_opt />
                <div class="lib-hero__ctas">
                    <button class="lib-hero__cta lib-hero__cta--primary" on:click=|_| scroll_to_grid()>
                        {book_icon()}
                        "Start reading"
                    </button>
                    <a class="lib-hero__cta lib-hero__cta--ghost" href="/blog">"Read the blog"</a>
                </div>
            </section>
            <ContinueCard index=index_opt />
            {move || match index.get() {
                AsyncResult::Loading => view! { <p class="muted">"Loading the library…"</p> }.into_any(),
                AsyncResult::Failed(message) => {
                    view! { <p class="error">"The library failed to load: " {message}</p> }.into_any()
                }
                AsyncResult::Loaded(idx) => book_grid(&idx).into_any(),
            }}
            <crate::shell::footer::SiteFooter />
        </div>
    }
}

/// Smooth-jump to the grid, offset for the sticky header.
fn scroll_to_grid() {
    let Some(document) = web_sys::window().and_then(|w| w.document()) else {
        return;
    };
    let Some(grid) = document.get_element_by_id("library-grid") else {
        return;
    };
    if let Some(window) = web_sys::window() {
        let top = grid.get_bounding_client_rect().top() + window.scroll_y().unwrap_or(0.0) - 80.0;
        window.scroll_to_with_x_and_y(0.0, top);
    }
}

fn book_grid(index: &SynapseIndexDto) -> impl IntoView + use<> {
    view! {
        <section id="library-grid" class="lib-section library__body">
            <h2 class="lib-section__title">"The books"</h2>
            <div class="lib-grid">{entries_view(&index.entries)}</div>
        </section>
    }
}

fn entries_view(entries: &[CatalogEntryDto]) -> Vec<AnyView> {
    entries
        .iter()
        .map(|entry| match entry {
            CatalogEntryDto::Category(category) => {
                let books: Vec<_> = category
                    .entries
                    .iter()
                    .filter_map(|e| match e {
                        CatalogEntryDto::Book(book) => Some(book_card(book)),
                        CatalogEntryDto::Category(_) => None,
                    })
                    .collect();
                view! {
                    <div class="lib-group">
                        <div class="lib-group__title">{category.title.clone()}</div>
                        <div class="lib-grid lib-grid--nested">{books}</div>
                    </div>
                }
                .into_any()
            }
            CatalogEntryDto::Book(book) => book_card(book),
        })
        .collect()
}

/// "N/M read" — shown only once there is something to report, so an untouched library stays
/// uncluttered and the chip always means "you have started this".
fn progress_chip(book: &BookDto) -> AnyView {
    let progress = crate::catalog::state::ProgressStore::from_context();
    let book = book.clone();
    view! {
        {move || {
            let done = progress.done().read();
            let count = logic::progress::completed_count(&book, &done);
            let total = logic::lesson_count(&book);
            (count > 0 && total > 0)
                .then(|| {
                    view! {
                        <span
                            class="lib-card__progress"
                            class:lib-card__progress--all=count == total
                        >
                            {format!("{count}/{total} read")}
                        </span>
                    }
                })
        }}
    }
    .into_any()
}

/// "Pick up where you left off" — the one thing that makes a second visit different from a
/// first. Renders nothing until there IS a last lesson, so a new reader never sees an empty
/// shelf, and nothing when the index has not loaded, because the title would be a raw slug.
#[component]
fn ContinueCard(index: Signal<Option<SynapseIndexDto>>) -> impl IntoView {
    let progress = state::ProgressStore::from_context();
    move || {
        let path = progress.last().get()?;
        let index = index.get()?;
        let segments: Vec<String> = path.split('/').map(str::to_owned).collect();
        // The title comes from the index rather than being stored alongside the path: a stored
        // title would go stale the moment a lesson is renamed, and the path is the only thing
        // that has to stay true.
        let book = logic::book_of(&index, &segments)?;
        let title = logic::reading_order(book)
            .into_iter()
            .find(|(p, _)| *p == path)
            .map(|(_, lesson)| lesson.title)?;
        Some(view! {
            <a class="lib-continue" href=format!("/synapse/{path}")>
                <span class="lib-continue__label">"Pick up where you left off"</span>
                <span class="lib-continue__title">{title}</span>
                <span class="lib-continue__book">{book.title.clone()}</span>
            </a>
        })
    }
}

fn book_card(book: &BookDto) -> AnyView {
    let chapters = logic::chapter_count(book);
    let lessons = logic::lesson_count(book);
    let mut meta: Vec<String> = Vec::new();
    if chapters > 0 {
        meta.push(format!("{chapters} {}", plural(chapters, "chapter")));
    }
    meta.push(format!("{lessons} {}", plural(lessons, "lesson")));
    if let Some(minutes) = book.estimated_reading_minutes {
        meta.push(format!("~{minutes} min"));
    }
    let meta = meta.join(" · ");
    let tags: Vec<_> = book
        .tags
        .iter()
        .take(3)
        .map(|t| view! { <span class="lib-card__tag">{t.clone()}</span> })
        .collect();
    let body = view! {
        <div class="lib-card__meta">{book_icon()}<span>{meta}</span></div>
        <div class="lib-card__title">{book.title.clone()}</div>
        {(!book.description.is_empty())
            .then(|| view! { <p class="lib-card__desc">{book.description.clone()}</p> })}
        <div class="lib-card__footer">
            {tags}
            {progress_chip(book)}
            <span class="lib-card__cta">"Read" {arrow_icon()}</span>
        </div>
    };
    match logic::first_lesson_path(book) {
        Some(path) => view! {
            <a class="lib-card" href=format!("/synapse/{path}")>{body}</a>
        }
        .into_any(),
        None => view! { <div class="lib-card lib-card--empty">{body}</div> }.into_any(),
    }
}

fn plural(n: usize, word: &str) -> String {
    if n == 1 {
        word.to_owned()
    } else {
        format!("{word}s")
    }
}

fn book_icon() -> impl IntoView {
    view! {
        <svg class="lib-card__meta-ic" viewBox="0 0 24 24" width="14" height="14" fill="none"
             stroke="currentColor" stroke-width="2" stroke-linecap="round"
             stroke-linejoin="round" aria-hidden="true">
            <path d="M2 3h6a4 4 0 0 1 4 4v14a3 3 0 0 0-3-3H2z"></path>
            <path d="M22 3h-6a4 4 0 0 0-4 4v14a3 3 0 0 1 3-3h7z"></path>
        </svg>
    }
}

fn arrow_icon() -> impl IntoView {
    view! {
        <svg class="lib-card__cta-ic" viewBox="0 0 24 24" width="14" height="14" fill="none"
             stroke="currentColor" stroke-width="2" stroke-linecap="round"
             stroke-linejoin="round" aria-hidden="true">
            <path d="M5 12h14 M13 6l6 6-6 6"></path>
        </svg>
    }
}

File: web/src/islands/library.ts
import * as log from "../lib/log";
// The library landing's post-hydration chrome. Vanilla TS, no framework — the SSR page is plain
// HTML and reading progress is a browser-only concern (`localStorage` has no SSR equivalent), so
// there is nothing here for a component framework to hydrate INTO; this script just mutates the
// DOM Astro already rendered.
//
// Three jobs:
//   (a) inject a progress bar + pct into each book card's footer (`.lib-card__progress` / `--all`),
//   (b) render the `.lib-continue` "pick up where you left off" card above the grid, and
//   (c) the "Start reading" CTA's smooth-scroll-with-header-offset (the grid's bounding-rect top
//       + `scrollY` − 80, the sticky header's height).
//
// ISLAND-PROPS PATTERN CHOSEN HERE: index.astro embeds the SAME `SynapseIndexDto` it already
// fetched for SSR as a `<script type="application/json" id="library-index-data">` blob
// (`<` escaped to `<` so a title containing "</script" can't break the tag out early).
// This island parses that blob ONCE and re-uses the PURE `tree.ts`/`progress.ts` helpers the
// SSR page already computed with — no second network round trip, and no duplicated
// flatten/lookup logic living twice in two languages of client code. Each card carries
// `data-book-slug` (see BookCard.astro) so this script can re-resolve a book without
// re-walking the whole catalog tree per card; the continue card resolves its book through
// `bookOf` the same way.
import * as storage from "../lib/storage";
import * as api from "../lib/api/client";
import { AUTH_CHANGED, isAuthed } from "./workbench/contracts";
import { completedCount, parse as parseDone } from "../lib/catalog/progress";
import { bookOf, findBook, readingOrder } from "../lib/catalog/tree";
import type { SynapseIndex } from "../lib/api/client";

function readIndex(): SynapseIndex | null {
  const el = document.getElementById("library-index-data");
  if (!el?.textContent) return null;
  try {
    return JSON.parse(el.textContent) as SynapseIndex;
  } catch {
    return null;
  }
}

/** (a) A card progress bar + pct — shown only once there is something to report, so an untouched
 *  library stays exactly as SSR rendered it. */
function injectProgressChips(index: SynapseIndex, done: Set<string>): void {
  if (done.size > 0) log.debug(`library: injecting progress chips (${done.size} finished lessons)`);
  for (const card of document.querySelectorAll<HTMLElement>("[data-book-slug]")) {
    const slug = card.dataset.bookSlug;
    if (!slug) continue;
    const book = findBook(index, slug);
    if (!book) continue;
    const total = readingOrder(book).length;
    const count = completedCount(book, done);

    const footer = card.querySelector(".lib-card__footer");
    const cta = footer?.querySelector(".lib-card__cta");
    if (!footer || !cta) continue;
    // Idempotent: a re-inject (after the signed-in server merge) replaces the device-only chip.
    footer.querySelector(".lib-card__progress")?.remove();
    if (count <= 0 || total <= 0) continue;

    const pct = Math.round((count / total) * 100);
    const chip = document.createElement("div");
    chip.className = count === total ? "lib-card__progress lib-card__progress--all" : "lib-card__progress";
    const bar = document.createElement("div");
    bar.className = "lib-card__progress-bar";
    const fill = document.createElement("div");
    fill.className = "lib-card__progress-fill";
    fill.style.setProperty("--pct", `${pct}%`);
    bar.append(fill);
    const label = document.createElement("span");
    label.className = "lib-card__progress-pct";
    label.textContent = count === total ? "Done" : `${pct}%`;
    chip.append(bar, label);
    footer.insertBefore(chip, cta);
  }
}

/** (b) "Pick up where you left off" — renders nothing until there IS a last lesson. The
 *  title/book come from the index rather than being stored alongside the path: a stored title
 *  would go stale the moment a lesson is renamed. */
function renderContinueCard(index: SynapseIndex): void {
  const mount = document.getElementById("lib-continue-mount");
  const last = storage.get(storage.READER_LAST_KEY);
  if (last != null) log.debug(`library: resume card → ${last}`);
  if (!mount) return;
  const path = storage.get(storage.READER_LAST_KEY);
  if (!path) return;

  const segments = path.split("/").filter((segment) => segment !== "");
  const book = bookOf(index, segments);
  if (!book) return;
  const entry = readingOrder(book).find((candidate) => candidate.path === path);
  if (!entry) return;

  const a = document.createElement("a");
  a.className = "lib-continue";
  a.href = `/synapse/${path}`;

  const label = document.createElement("span");
  label.className = "lib-continue__label";
  label.textContent = "Pick up where you left off";

  const title = document.createElement("span");
  title.className = "lib-continue__title";
  title.textContent = entry.lesson.title;

  const bookLine = document.createElement("span");
  bookLine.className = "lib-continue__book";
  bookLine.textContent = book.title;

  a.append(label, title, bookLine);
  mount.replaceChildren(a);
}

/** (c) Smooth-jump to the grid, offset for the sticky header. */
function wireStartReading(): void {
  document.getElementById("lib-start-reading")?.addEventListener("click", () => {
    const grid = document.getElementById("library-grid");
    if (!grid) return;
    const top = grid.getBoundingClientRect().top + window.scrollY - 80;
    window.scrollTo(0, top);
  });
}

/** Merge the signed-in reader's account progress into the chips (now or when auth adopts late).
 *  Anonymous → `listProgress` returns `[]`, a no-op; the device chips already showed. */
function syncChipsFromServer(index: SynapseIndex, done: Set<string>): void {
  if (!isAuthed()) return;
  void api
    .listProgress()
    .then((server) => {
      let added = false;
      for (const path of server) {
        if (!done.has(path)) {
          done.add(path);
          added = true;
        }
      }
      if (added) injectProgressChips(index, done);
    })
    .catch((error) => log.debug(`library progress sync skipped: ${error instanceof Error ? error.message : String(error)}`));
}

function init(): void {
  wireStartReading();
  const index = readIndex();
  if (!index) return;
  const done = parseDone(storage.get(storage.READER_PROGRESS_KEY));
  injectProgressChips(index, done);
  renderContinueCard(index);
  syncChipsFromServer(index, done);
  window.addEventListener(AUTH_CHANGED, () => syncChipsFromServer(index, done));
}

if (document.readyState === "loading") {
  document.addEventListener("DOMContentLoaded", init);
} else {
  init();
}
